### PR TITLE
Containers get new mounts for local DPs

### LINKED ISSUE #542: Tasks using Docker cannot use files on local data providers
Task using Docker currently mount the following directories through `-v` options in the `docker run` command:
* present working directory (a.k.a task execution directory).
* bourreau dp cache, so that synchronized files are accessible from the container.
* task directory, so that symbolic link "DP_Cache" (that points to the bourreau dp cache) can be resolved in the container.

However, in case of a local data provider, the files in the execution directory are symlinked directly to the data provider directory which is none of the above. Thus, the symlnks to these files are broken. 

To fix this issue, one should:
* Test if the task input files are located on a local data provider.
* If yes, add the base path of this data provider to the list of directories that will be mounted in the container.

### PR DESCRIPTION
**Pull request by prioux**
Containers environments (singularity and docker)
all get explicit mounts for local DPs.

This will fix #542 and #692 .

This has not been actually tested, sorry! At least I verified that the code seems to produce the proper strings, using the console. Please try it out.

### CODE REVIEW

**Comment by prioux** [comment]
<farnsworth>Good news everyone!</farnsworth> I tested this with a singularity container and the mounts worked perfectly. Now if someone could test with docker?

**Comment by prioux** [comment]
Even better news: it works perfectly fine in Docker too.

**Comment by gkiar** [comment]
You're a rockstar!

**Comment by shots47s** [comment]
Wonderful!

On Feb 17, 2018 1:48 PM, "Greg Kiar" <notifications@github.com> wrote:

You're a rockstar!

—
You are receiving this because you were assigned.
Reply to this email directly, view it on GitHub
<https://github.com/aces/cbrain/pull/707#issuecomment-366462690>, or mute
the thread

File: BrainPortal/app/models/cluster_task.rb [hunk 1]
```diff
@@ -2004,6 +2004,12 @@ def docker_commands(command_script)
     # 5) Basename of the docker wrapper script
     docker_wrapper_basename = ".dockerjob.#{self.run_id}.sh"
 
+    # 6) More -v (bind mounts) for all the local data providers.
+    # This will be a string "-v path1:path1 -v path2:path2" etc.
+    esc_local_dp_mountpoints = local_dp_storage_paths.inject("") do |dock_opts,path|
```
[the remark this answers is not in the document]
Comment (prioux) [reply BrainPortal/app/models/cluster_task.rb]: I'm OK with the blank line; as long as it ends with a backslash, bash will just intepret it as whitespace between arguments.

**Pull request merged**